Code/Framework/AzCore/AzCore/Memory/AllocatorBase.cpp

```diff
@@ -6,8 +6,155 @@
  *
  */
 
-#include <AzCore/Memory/AllocatorManager.h>
 #include <AzCore/Memory/Memory.h>
+#include <AzCore/Memory/AllocatorManager.h>
+
+#define RECORDING_ENABLED 0
+
+#if RECORDING_ENABLED
+
+#include <AzCore/std/containers/map.h>
+#include <AzCore/IO/SystemFile.h>
+#include <AzCore/std/parallel/mutex.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+
+namespace
+{
+    class DebugAllocator
+    {
+    public:
+        typedef void* pointer_type;
+        typedef AZStd::size_t size_type;
+        typedef AZStd::ptrdiff_t difference_type;
+        typedef AZStd::false_type allow_memory_leaks; ///< Regular allocators should not leak.
+
+        AZ_FORCE_INLINE pointer_type allocate(size_t byteSize, size_t alignment, int = 0)
+        {
+            return AZ_OS_MALLOC(byteSize, alignment);
+        }
+        AZ_FORCE_INLINE size_type resize(pointer_type, size_type)
+        {
+            return 0;
+        }
+        AZ_FORCE_INLINE void deallocate(pointer_type ptr, size_type, size_type)
+        {
+            AZ_OS_FREE(ptr);
+        }
+    };
+
+    #pragma pack(push, 1)
+    struct alignas(1) AllocatorOperation
+    {
+        enum OperationType : size_t
+        {
+            ALLOCATE,
+            DEALLOCATE
+        };
+        OperationType m_type: 1;
+        size_t m_size : 28; // Can represent up to 256Mb requests
+        size_t m_alignment : 7; // Can represent up to 128 alignment
+        size_t m_recordId : 28; // Can represent up to 256M simultaneous requests, we reuse ids
+    };
+    #pragma pack(pop)
+    static_assert(sizeof(AllocatorOperation) == 8);
+
+    static AZStd::mutex s_operationsMutex = {};
+
+    static constexpr size_t s_maxNumberOfAllocationsToRecord = 16384;
+    static size_t s_numberOfAllocationsRecorded = 0;
+    static constexpr size_t s_allocationOperationCount = 5 * 1024;
+    static AZStd::array<AllocatorOperation, s_allocationOperationCount> s_operations = {};
+    static uint64_t s_operationCounter = 0;
+
+    static unsigned int s_nextRecordId = 1;
+    using AllocatorOperationByAddress = AZStd::map<void*, AllocatorOperation, AZStd::less<void*>, DebugAllocator>;
```

Comment:
Do we need the ordered traversal guarantee of a `map` or could we use the more performant `unordered_map` instead?

Reply:
we can use `unordered_map`, is just that the the usage of this map was left outside of the benchmark measurements and this container can receive a lot of items. I will change it to unordered and see how it behaves.